doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md

```diff
@@ -0,0 +1,607 @@
+print("Saved PyTorch Model State to model.pth")
+```
+
+We'll cover the rest of the tutorial (loading the model and doing batch prediction) later!
+
++++
+
+## Introducing a wrapper function (no Ray AIR, yet!)
+The notebook-style from the tutorial is great for tutorials, but in your production code you probably wrapped the actual training logic in a function. So let's do this here, too.
+
+Note that we do not add or alter any code here (apart from variable definitions) - we just take the loose bits of code in the current tutorial and put them into one function.
+
+```{code-cell} ipython3
+def train_func():
+    batch_size = 64
+    lr = 1e-3
+    epochs = 5
+    
+    # Create data loaders.
+    train_dataloader = DataLoader(training_data, batch_size=batch_size)
+    test_dataloader = DataLoader(test_data, batch_size=batch_size)
+    
+    # Get cpu or gpu device for training.
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    print(f"Using {device} device")
+    
+    model = NeuralNetwork().to(device)
+    print(model)
+    
+    loss_fn = nn.CrossEntropyLoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=lr)
+    
+    for t in range(epochs):
+        print(f"Epoch {t+1}\n-------------------------------")
+        train_epoch(train_dataloader, model, loss_fn, optimizer)
+        test_epoch(test_dataloader, model, loss_fn)
+
+    print("Done!")
+```
+
+Let's see it in action again:
+
+```{code-cell} ipython3
+train_func()
+```
+
+The output should look very similar to the previous ouput.
+
++++
+
+## Starting with Ray AIR: Distribute the training
+
+As a first step, we want to distribute the training across multiple workers. For this we want to
+
+1. Use data-parallel training by sharding the training data
+2. Setup the model to communicate gradient updates across machines
+3. Report the results back to Ray Train.
+
+
+To facilitate this, we only need a few changes to the code:
+
+1. We import Ray Train:
+
+```python
+import ray.train as train
+```
+
+
+2. We use a `config` dict to configure some hyperparameters (this is not strictly needed but good practice):
+
+```python
+def train_func(config: dict):
+    batch_size = config["batch_size"]
+    lr = config["lr"]
+    epochs = config["epochs"]
+```
+
+3. We dynamically adjust the worker batch size according to the number of workers:
+
+```python
+    batch_size_per_worker = batch_size // train.world_size()
+```
+
+4. We prepare the data loader for distributed data sharding:
+
+```python
+    train_dataloader = train.torch.prepare_data_loader(train_dataloader)
+    test_dataloader = train.torch.prepare_data_loader(test_dataloader)
+```
+
+5. We prepare the model for distributed gradient updates:
+
+```python
+    model = train.torch.prepare_model(model)
+```
+
+Note that `train.torch.prepare_model()` also automatically takes care of setting up devices (e.g. GPU training) - so we can get rid of those lines in our current code!
```

Comment:
Should we explain that this wraps the model in DDP?

Reply:
I think we shouldn't be too verbose here, it's not about technical detail but more for hands on.

Reply:
That's fair, but people may end up wrapping the model themselves if they don't know it will already be done IMO